lib/portage/dep/dep_check.py

```diff
@@ -694,15 +681,35 @@ def mydbapi_match_pkgs(atom):
 			# contribute to outcomes that appear to be random. Meanwhile,
 			# the order specified in the ebuild is without variance, so it
 			# does not have this problem.
-			sort_keys.append(lambda x: x.new_slot_count)
+			choices.sort(key=operator.attrgetter('new_slot_count'))
 
-		choices.sort(key=lambda x: tuple(f(x) for f in sort_keys))
 		for choice_1 in choices[1:]:
 			cps = set(choice_1.cp_map)
 			for choice_2 in choices:
 				if choice_1 is choice_2:
 					# choice_1 will not be promoted, so move on
 					break
+				if (
+					# For removal actions, prefer choices where all packages
+					# have been pulled into the graph.
+					(graph_interface and graph_interface.removal_action and
+					choice_1.all_in_graph and not choice_2.all_in_graph)
```

Comment:
This must be the case that triggers that triggers removal of the latest python slot in https://bugs.gentoo.org/707108. We should be able to correct the problem be using the all_installed_slots attribute, for symmetry with the action that installs the new python slot.

Reply:
oh? if I understand this correctly, python won't be removed after the fix?
But I was under the impression that the issue was that python 3_7 was wrongly emerged(because all(?) python version from PYTHON_COMPAT were emerged), and correctly depcleaned, because without this commit, it wouldn't be emerged at all by portage.

Reply:
> oh? if I understand this correctly, python won't be removed after the fix?

Correct.

> But I was under the impression that the issue was that python 3_7 was wrongly emerged(because all(?) python versions from PYTHON_COMPAT were emerged), and correctly depcleaned, because without this commit, it wouldn't be emerged at all by portage.

Your impression is the same as the one expressed in https://bugs.gentoo.org/705700. Please refer to my response in https://bugs.gentoo.org/705700#c12 (and https://bugs.gentoo.org/706278 is an excellent example of what I'm talking about).

Reply:
Understood. Thank you for clarifying.

I was confused because only the first of the following 3 emerges did bring in that python:
```
# emerge --deep --update --with-bdeps=y @world
# emerge --deep --update --with-bdeps=n @world
# emerge --update --with-bdeps=y @world
```
Unless I'm misunderstanding how they work(which is quite likely), after the fix, all 3 emerges above will bring python in?

Reply:
Those 3 emerge commands are all have very important differences which make it impossible to guarantee that they'll all bring in the new python slot. However, the first one is guaranteed to pull it in.

Reply:
https://github.com/gentoo/portage/pull/502 has a fix.